### ISSUE TITLE
Support Postgres-style variable substitution

### ISSUE THREAD

**Issue by benji-york** [sqlfluff/sqlfluff#3343]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

The Postgres `psql` utility supports flavor of colon-style variable substitution that currently confuses sqlfluff.  E.g.,

```sql
ALTER TABLE name:variable RENAME TO name;
```

Running the above through sqlfluff produces this output:

```
sqlfluff lint --dialect postgres 2.sql
== [2.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'ALTER
                       | TABLE name:variable RENAME TO name...'
```

### Use case

I would like it if in the above the string "name:variable" were considered a valid table name (and other identifiers similarly).

### Dialect

This applies to the Postgres dialect.

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
This sounds more like a templater feature than a dialect feature. Does psql allow variables to contain SQL fragments, e.g.: `WHERE foo = '3'`?

**Comment by benji-york** [comment]
> This sounds more like a templater feature than a dialect feature.

True!  After looking over the code some, that may well be the right place to implement this.

> Does psql allow variables to contain SQL fragments, e.g.: WHERE foo = '3'?

Yes.  E.g.,

```
% psql -v expression='2 + 2'
psql (14.2, server 10.18)
SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)
Type "help" for help.

db=> select :expression;
 ?column?
----------
        4
(1 row)

db=> select 5:expression;
 ?column?
----------
       54
(1 row)
```

More at the [docs](https://www.postgresql.org/docs/current/app-psql.html#APP-PSQL-VARIABLES).

**Comment by tunetheweb** [comment]
Can we not do this already using the following config:

```
[sqlfluff]
templater = placeholder

[sqlfluff:templater:placeholder]
param_style = colon
expression = '2 + 2'
```

When I have that as my config file, it parses correctly:

```
% echo "select :expression;" | sqlfluff parse --dialect ansi -
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    literal:                                  "'2 + 2'"
[L:  1, P: 19]      |            [META] dedent:
[L:  1, P: 19]      |    statement_terminator:                                     ';'
[L:  1, P: 20]      |    newline:                                                  '\n'
```

**Comment by benji-york** [comment]
> Can we not do this already using the following config:
> [snip]

It doesn't seem so:

```
% cat .sqlfluff
[sqlfluff]
templater = placeholder

[sqlfluff:templater:placeholder]
param_style = colon
expression = '20201225'
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
% cat 3.sql
select :expression
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
% tox-ve/bin/sqlfluff lint --dialect postgres 3.sql
All Finished 📜 🎉!
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
% cat 2.sql
ALTER TABLE name:expression RENAME TO name;
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
% tox-ve/bin/sqlfluff lint --dialect postgres 2.sql
== [2.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'ALTER
                       | TABLE name:expression RENAME TO na...'
All Finished 📜 🎉!
```

**Comment by tunetheweb** [comment]
Can you explain that `name:expression` thing? Should that evaluate to `name20201225`?

**Comment by tunetheweb** [comment]
BTW confirmed that won't work:

https://github.com/sqlfluff/sqlfluff/blob/ae5c5ef669a9aff4fb394b01f4d615f360ac7d9f/src/sqlfluff/core/templaters/placeholder.py#L23

So currently it doesn't allow the character before the colon to be a colon (to avoid cast like `SELECT col1::int`), a word character (`\w` - so `[A-Za-z0-9]`) or the backslash character (`\x5c`).

So basically it needs to be surrounded by whitespaces. I'm guessing this is to avoid matching false positives (e.g. if you have a placeholder called `:int`.

We could either relax that requirement and accept the false positives. Or perhaps introduce a. new type (`colon_all`?) with this relaxed?

**Comment by benji-york** [comment]
> Can you explain that `name:expression` thing? Should that evaluate to `name20201225`?

Yep!  In one of my use cases it is used to add a date suffix to table names for an ETL.

**Comment by tunetheweb** [comment]
I've added `colon_nospaces` templating type to #3354

I did add this warning though:

```
    -- colon_nospaces
    -- (use with caution as more prone to false positives)
    WHERE bla = table:my_name
```

**Comment by tunetheweb** [comment]
Closed with #3354

**Comment by benji-york** [comment]
Thanks!  So far `colon_nospaces` seems to be working for me in limited testing.
